library/lua/profiler.lua

```diff
@@ -0,0 +1,651 @@
+    else
+      parent_stats.name_child_time = parent_stats.name_child_time + time
+    end
+    local ch = parent_stats.children[func]
+    if not ch then
+      parent_stats.children[func] = 1
+      parent_stats.children_time[func] = time
+    else
+      parent_stats.children[func] = ch + 1
+      parent_stats.children_time[func] = parent_stats.children_time[func] + time
+    end
+    parent.profile_time = parent.profile_time + this.profile_time + (os.clock() - entry)
+  end
+end
+
+local function _profiler_hook_wrapper_by_time()
+  local self = _profiler.running
+  local timetaken = os.clock() - self.lastclock
+  local rawstats = self.rawstats
+  local prevented = self.prevented_functions
+  local depth = 2
+  local caller = debug.getinfo(depth,'fl')
+  local child
+  if not caller then
+    return
+  end
+  local cf = caller.func
+  if cf then
+    child = get_stats(rawstats, prevented, cf, depth)
+    if child then
+      child.count = child.count + 1
+      child.time = child.time + timetaken
+      local line = caller.currentline
+      child.currentline[line] = (child.currentline[line] or 0) + 1
+    else
+      cf = DEFAULT_FILTERED_FUNC
+    end
+  else
+    cf = DEFAULT_MISSING_FUNC
+  end
+  depth = 3
+  local caller = debug.getinfo(depth,'f')
+  while caller do
+    if caller.func then
+      local this = get_stats(rawstats, prevented, caller.func, depth)
+      if this then
+        this.time = this.time + timetaken
+        if not child or child.anon then
+          this.anon_child_time = this.anon_child_time + timetaken
+        else
+          this.name_child_time = this.name_child_time + timetaken
+        end
+        local ch = this.children[cf]
+        if ch then
+          this.children[cf] = ch + 1
+          this.children_time[cf] = this.children_time[cf] + timetaken
+        else
+          this.children[cf] = 1
+          this.children_time[cf] = timetaken
+        end
+        cf = caller.func
+      else
+        cf = DEFAULT_FILTERED_FUNC
+      end
+      child = this
+    else
+      cf = DEFAULT_MISSING_FUNC
+      child = nil
+    end
+    depth = depth + 1
+    caller = debug.getinfo(depth, 'f')
+  end
+  self.lastclock = os.clock()
+end
+
+
+--
+-- This function starts the profiler.  It will do nothing
+-- if this (or any other) profiler is already running.
+--
+function _profiler.start(self)
+  if _profiler.running then
+    return
+  end
+  -- Start the profiler. This begins by setting up internal profiler state
+  _profiler.running = self
+  assert(_profiler.running)
+  self.rawstats = {}
+  self.stack = {}
+  self.depth = 0
+  if self.variant == "time" then
+    self.lastclock = os.clock()
+    debug.sethook( _profiler_hook_wrapper_by_time, "", self.sampledelay )
+  elseif self.variant == "call" then
+    debug.sethook( _profiler_hook_wrapper_by_call, "cr" )
+  else
+    print("Profiler method must be 'time' or 'call'.")
+    sys.exit(1)
```

Comment:
Not acceptable in a module (but I don't think this is defined, anyway). It should use error() instead.